pandas/core/array_algos/putmask.py

```diff
@@ -114,7 +114,9 @@ def putmask_smart(values: np.ndarray, mask: np.ndarray, new) -> np.ndarray:
 
     new = np.asarray(new)
 
-    if values.dtype.kind == new.dtype.kind:
+    values_kind = values.dtype.kind
+    new_kind = new.dtype.kind
+    if values_kind == new_kind or (values_kind == "f" and new_kind in ["i", "u"]):
         # preserves dtype if possible
         return _putmask_preserve(values, new, mask)
```

Comment:
The added case here does not affect the `fillna` behavior, but is required to maintain the behavior that setting a float series with a sequence of ints preserves the float type (covered in  `test_where_unsafe_float`)
 
Based on the comment here, this change seemed appropriate since settings ints into floats should be able to preserve dtype.